go-controller/pkg/ovn/secondary_layer2_network_controller.go

```diff
@@ -0,0 +1,298 @@
+	if err := oc.Init(); err != nil {
+		return err
+	}
+
+	return oc.Run()
+}
+
+// Stop gracefully stops the controller, and delete all logical entities for this network if requested
+func (oc *SecondaryLayer2NetworkController) Stop() {
+	klog.Infof("Stop secondary %s network controller of network %s", oc.TopologyType(), oc.GetNetworkName())
+	close(oc.stopChan)
+	oc.wg.Wait()
+
+	if oc.podHandler != nil {
+		oc.watchFactory.RemovePodHandler(oc.podHandler)
+	}
+}
+
+// Cleanup cleans up logical entities for the given network, called from net-attach-def routine
+// could be called from a dummy Controller (only has CommonNetworkControllerInfo set)
+func (oc *SecondaryLayer2NetworkController) Cleanup(netName string) error {
+	klog.Infof("Delete OVN logical entities for %s network controller of network %s", types.Layer2Topology, netName)
+
+	// delete layer 2 logical switches
+	ops, err := libovsdbops.DeleteLogicalSwitchesWithPredicateOps(oc.nbClient, nil,
+		func(item *nbdb.LogicalSwitch) bool {
+			return item.ExternalIDs[types.NetworkExternalID] == netName
+		})
+	if err != nil {
+		return fmt.Errorf("failed to get ops for deleting switches of network %s: %v", netName, err)
+	}
+
+	_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
+	if err != nil {
+		return fmt.Errorf("failed to deleting switches of network %s: %v", netName, err)
+	}
+
+	return nil
+}
+
+func (oc *SecondaryLayer2NetworkController) Run() error {
+	klog.Infof("Starting all the Watchers for network %s ...", oc.GetNetworkName())
+	start := time.Now()
+
+	if err := oc.WatchPods(); err != nil {
+		return err
+	}
+
+	klog.Infof("Completing all the Watchers for network %s took %v", oc.GetNetworkName(), time.Since(start))
+
+	// controller is fully running and resource handlers have synced, update Topology version in OVN
+	if err := oc.updateL2TopologyVersion(); err != nil {
+		return fmt.Errorf("failed to update topology version for network %s: %v", oc.GetNetworkName(), err)
+	}
+
+	return nil
+}
+
+func (oc *SecondaryLayer2NetworkController) Init() error {
+	switchName := oc.GetNetworkScopedName(types.OVNLayer2Switch)
+	logicalSwitch := nbdb.LogicalSwitch{
+		Name:        switchName,
+		ExternalIDs: map[string]string{},
+	}
+	if oc.IsSecondary() {
+		logicalSwitch.ExternalIDs[types.NetworkExternalID] = oc.GetNetworkName()
+		logicalSwitch.ExternalIDs[types.TopologyExternalID] = oc.TopologyType()
+	}
+
+	layer2NetConfInfo := oc.NetConfInfo.(*util.Layer2NetConfInfo)
+
+	hostSubnets := make([]*net.IPNet, 0, len(layer2NetConfInfo.ClusterSubnets))
+	for _, subnet := range layer2NetConfInfo.ClusterSubnets {
+		hostSubnets = append(hostSubnets, subnet)
+		if utilnet.IsIPv6CIDR(subnet) {
+			logicalSwitch.OtherConfig = map[string]string{"ipv6_prefix": subnet.IP.String()}
+		} else {
+			logicalSwitch.OtherConfig = map[string]string{"subnet": subnet.String()}
+		}
+	}
+
+	err := libovsdbops.CreateOrUpdateLogicalSwitch(oc.nbClient, &logicalSwitch, &logicalSwitch.OtherConfig)
```

Comment:
Or perhaps we need to set the external ids on a separate transaction to avoid overriding the topology-version?

Reply:
seems so. but wouldn't that be a problem for createOvnClusterRouter() function as well? setting "k8s-cluster-router" external-ids would overriding the topology-version?

Reply:
Indeed, that's something we should look into. We might be constantly attempting to upgrade the topology on restart which hopefully results in a noop if we are trying to set things as they actually are.

Reply:
it is mostly ok, as getting topology version and upgrade is mostly done before createOvnClusterRouter() is called. The only exception is ovnTopologyCleanup()->NonDualStackAddressSetCleanup().

Reply:
Right, a topology upgrade is a two step process, there is stuff that needs to happen at init and then there is stuff that needs to happen after we are actually running, as to not disrupt networking and the NonDualStackAddressSetCleanup is one of those things. For this specifically, aside for not cleaning up those address sets, it should not be of much consequence as those address sets would not be used anyway.

Let's try to come up with an idea to solve this here and use that to solve it for the router in a different PR?

Reply:
I've done as you suggested, to createSwitch and updateExternalDs in two different trasactions, so that the topology version external-ids is not overriden.

Reply:
Hmm I am thinking that the process that would look more consolidated where there would actually be a need to handle a topology upgrade would probably be:

1. attempt to read the original topology version and save that in a controller variable
2. as part of the normal flow, update the switch/router,  including the original topology that was read (as to override with the same value); and handle whatever extra actions the topology upgrade requires comparing that original topology version with the final topology version desired
3. update the switch/router with the final topology version

Reply:
I created this issue with the same proposal to fix it for the primary and L3 networks as well:
https://github.com/ovn-org/ovn-kubernetes/issues/3368